build_tools/cmake/iree_e2e_matmul_test.cmake

```diff
@@ -157,38 +164,81 @@ function(iree_single_backend_e2e_matmul_test)
     ${ARGN}
   )
 
-  iree_is_bytecode_module_test_excluded_by_labels(_EXCLUDED_BY_LABELS "${_RULE_LABELS}")
-  if(_EXCLUDED_BY_LABELS)
-    return()
+  # ---------------------------------------------------------------------------
+  # Bytecode module builds require
+  #   1. the compiler, either in the same build or provided in IREE_HOST_BIN_DIR
+  #   2. compiler support for _RULE_INPUT_TYPE
+  #   3. compiler support for _RULE_TARGET_BACKEND
+  set(_BYTECODE_MODULE_BUILD_ENABLED TRUE)
+
+  # 1. Check for the compiler.
+  if(NOT IREE_BUILD_COMPILER AND NOT IREE_HOST_BIN_DIR)
+    set(_BYTECODE_MODULE_BUILD_ENABLED FALSE)
   endif()
 
-  # Omit tests for which the specified driver or target backend is not enabled.
-  # This overlaps with directory exclusions and other filtering mechanisms.
-  string(TOUPPER ${_RULE_DRIVER} _UPPERCASE_DRIVER)
-  string(REPLACE "-" "_" _NORMALIZED_DRIVER ${_UPPERCASE_DRIVER})
-  if(NOT DEFINED IREE_HAL_DRIVER_${_NORMALIZED_DRIVER})
-    message(SEND_ERROR "Unknown driver '${_RULE_DRIVER}'. Check IREE_HAL_DRIVER_* options.")
+  # 2. Check target backend availability.
+  # Note: we can only reliably check for this when building the compiler host
+  # tools from source. If the tools are already built, we assume that all target
+  # backends are enabled. We could query the tools in the binary directory for
+  # support dynamically if optionality would be useful.
+  if(NOT IREE_HOST_BIN_DIR)
+    string(TOUPPER ${_RULE_TARGET_BACKEND} _UPPERCASE_TARGET_BACKEND)
+    string(REPLACE "-" "_" _NORMALIZED_TARGET_BACKEND ${_UPPERCASE_TARGET_BACKEND})
+    # TODO(scotttodd): allow plugins to provide external backends here
+    if(NOT DEFINED IREE_TARGET_BACKEND_${_NORMALIZED_TARGET_BACKEND})
+      message(SEND_ERROR "Unknown backend '${_RULE_TARGET_BACKEND}'. Check IREE_TARGET_BACKEND_* options.")
+    endif()
```

Comment:
Down at this level though - I'd like to remove the checks from the `.cmake` helper files entirely. Leaf test suites would then only be included if all conditions are met, like this: https://github.com/openxla/iree/blob/783adb5510f2ea540657c64b25e1abb23742142b/runtime/src/iree/hal/drivers/CMakeLists.txt#L139-L142

An alternate approach would be to let plugins set some CMake options that these `.cmake` files can look for. I'd rather be explicit though - implicit is easy to accidently misconfigure or require changes that span multiple repositories.

(no action needed here, just another place to update when doing that)